File: agentcp_node/src/group/cursor_store.ts
/**
 * Cursor store for local persistence of group message/event cursors.
 * Mirrors Python SDK: agentcp/group/cursor_store.py
 */

import { isNodeEnvironment } from '../websocket';
import { logger } from '../utils';

/**
 * Abstract interface for cursor persistence.
 */
export interface CursorStore {
    saveMsgCursor(groupId: string, msgCursor: number): void;
    saveEventCursor(groupId: string, eventCursor: number): void;
    loadCursor(groupId: string): [number, number]; // [msg_cursor, event_cursor]
    removeCursor(groupId: string): void;
    flush(): void;
    close(): void;
}

/**
 * In-memory + JSON file cursor store.
 *
 * Monotonic cursor advancement (never goes backward).
 * file_path="" means pure in-memory mode.
 * Node.js environment uses fs for file persistence; browser is pure in-memory.
 */
export class LocalCursorStore implements CursorStore {
    private _cursors: Record<string, { msg_cursor: number; event_cursor: number }> = {};
    private _filePath: string;
    private _dirty: boolean = false;

    constructor(filePath: string = "") {
        this._filePath = filePath;
        if (filePath && isNodeEnvironment) {
            this._load();
        }
    }

    saveMsgCursor(groupId: string, msgCursor: number): void {
        if (!this._cursors[groupId]) {
            this._cursors[groupId] = { msg_cursor: 0, event_cursor: 0 };
        }
        if (msgCursor > this._cursors[groupId].msg_cursor) {
            this._cursors[groupId].msg_cursor = msgCursor;
            this._dirty = true;
        }
    }

    saveEventCursor(groupId: string, eventCursor: number): void {
        if (!this._cursors[groupId]) {
            this._cursors[groupId] = { msg_cursor: 0, event_cursor: 0 };
        }
        if (eventCursor > this._cursors[groupId].event_cursor) {
            this._cursors[groupId].event_cursor = eventCursor;
            this._dirty = true;
        }
    }

    loadCursor(groupId: string): [number, number] {
        const entry = this._cursors[groupId];
        if (!entry) {
            return [0, 0];
        }
        return [entry.msg_cursor, entry.event_cursor];
    }

    removeCursor(groupId: string): void {
        if (groupId in this._cursors) {
            delete this._cursors[groupId];
            this._dirty = true;
        }
    }

    flush(): void {
        if (!this._filePath || !isNodeEnvironment) {
            return;
        }
        if (!this._dirty) {
            return;
        }
        this._write();
    }

    close(): void {
        this.flush();
    }

    private _write(): void {
        try {
            const fs = require('fs');
            fs.writeFileSync(this._filePath, JSON.stringify(this._cursors, null, 2), 'utf-8');
            this._dirty = false;
        } catch (e) {
            logger.error(`[CursorStore] write to ${this._filePath} failed:`, e);
        }
    }

    private _load(): void {
        try {
            const fs = require('fs');
            if (!fs.existsSync(this._filePath)) {
                return;
            }
            const content = fs.readFileSync(this._filePath, 'utf-8');
            if (content) {
                this._cursors = JSON.parse(content);
            }
        } catch (e) {
            console.debug(`[CursorStore] load from ${this._filePath} failed:`, e);
        }
    }
}

File: agentcp_c++_sdk/core/src/group/cursor_store.cpp
#include "agentcp/cursor_store.h"

#include <fstream>
#include <sstream>

// Use nlohmann/json for serialization
#include "third_party/json.hpp"
using json = nlohmann::json;

namespace agentcp {
namespace group {

// ============================================================
// LocalCursorStore
// ============================================================

LocalCursorStore::LocalCursorStore(const std::string& file_path)
    : file_path_(file_path) {
    if (!file_path_.empty()) {
        Load();
    }
}

LocalCursorStore::~LocalCursorStore() {
    Close();
}

void LocalCursorStore::SaveMsgCursor(const std::string& group_id, int64_t msg_cursor) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto& entry = cursors_[group_id];
    if (msg_cursor > entry.msg_cursor) {
        entry.msg_cursor = msg_cursor;
        dirty_ = true;
    }
}

void LocalCursorStore::SaveEventCursor(const std::string& group_id, int64_t event_cursor) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto& entry = cursors_[group_id];
    if (event_cursor > entry.event_cursor) {
        entry.event_cursor = event_cursor;
        dirty_ = true;
    }
}

std::pair<int64_t, int64_t> LocalCursorStore::LoadCursor(const std::string& group_id) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = cursors_.find(group_id);
    if (it == cursors_.end()) {
        return {0, 0};
    }
    return {it->second.msg_cursor, it->second.event_cursor};
}

void LocalCursorStore::RemoveCursor(const std::string& group_id) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = cursors_.find(group_id);
    if (it != cursors_.end()) {
        cursors_.erase(it);
        dirty_ = true;
    }
}

void LocalCursorStore::Flush() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (file_path_.empty() || !dirty_) return;
    Write();
}

void LocalCursorStore::Close() {
    Flush();
}

void LocalCursorStore::Write() {
    try {
        json j = json::object();
        for (auto& [gid, entry] : cursors_) {
            j[gid] = {
                {"msg_cursor", entry.msg_cursor},
                {"event_cursor", entry.event_cursor}
            };
        }
        std::ofstream ofs(file_path_);
        if (ofs.is_open()) {
            ofs << j.dump(2);
            dirty_ = false;
        }
    } catch (...) {
        // silently ignore write errors
    }
}

void LocalCursorStore::Load() {
    try {
        std::ifstream ifs(file_path_);
        if (!ifs.is_open()) return;

        std::ostringstream ss;
        ss << ifs.rdbuf();
        std::string content = ss.str();
        if (content.empty()) return;

        auto j = json::parse(content);
        for (auto& [key, val] : j.items()) {
            CursorEntry entry;
            entry.msg_cursor = val.value("msg_cursor", (int64_t)0);
            entry.event_cursor = val.value("event_cursor", (int64_t)0);
            cursors_[key] = entry;
        }
    } catch (...) {
        // silently ignore load errors
    }
}

}  // namespace group
}  // namespace agentcp
